### PR TITLE
Add partial typing to ``class_checker``

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

### CODE REVIEW

File: pylint/checkers/classes/class_checker.py [hunk 1]
```diff
@@ -702,17 +711,19 @@ def _has_same_layout_slots(slots, assigned_value):
 }
 
 
-def _scope_default():
-    return collections.defaultdict(list)
+def _scope_default() -> defaultdict[str, list[_AccessNodes]]:
+    return defaultdict(list)
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:715]: Should we remove this function ?

File: pylint/checkers/classes/class_checker.py [hunk 2]
```diff
@@ -1930,7 +1943,14 @@ def _check_first_arg_for_type(self, node, metaclass=0):
...
             self.add_message("no-self-argument", node=node)
 
-    def _check_first_arg_config(self, first, config, node, message, method_name):
+    def _check_first_arg_config(
+        self,
+        first,
+        config: Sequence[str],
+        node: nodes.FunctionDef,
+        message: str,
+        method_name,
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1952]:
```suggestion
        method_name: str,
```

File: pylint/checkers/classes/class_checker.py [hunk 3]
```diff
@@ -1930,7 +1943,14 @@ def _check_first_arg_for_type(self, node, metaclass=0):
...
+    ):
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1953]:
```suggestion
    ) -> None:
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: pylint/checkers/classes/class_checker.py (hunk 1 above)
Comment (DanielNoord) [reply pylint/checkers/classes/class_checker.py:715]: We can't. Apparently you can't nest `defaultdicts`... Which is probably why this was done like this in the first place.

Comment (Pierre-Sassoulas) [reply pylint/checkers/classes/class_checker.py:715]: All right ! 

```suggestion
def _scope_default() -> defaultdict[str, list[_AccessNodes]]:
    # It's impossible to nest defaultdicts so we must use a function
    return defaultdict(list)
```

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**